File: typescript/src/Billing/BillingManager.ts
/**
 * Zirve Billing Manager — Lago Integration.
 *
 * Lightweight wrapper around Lago HTTP API for usage-based billing.
 */
export class BillingManager {
  private config: Record<string, any>;
  private url: string;
  private apiKey: string;

  constructor(config: Record<string, any>) {
    this.config = config;
    this.url = (config.url || 'http://lago-api.zirve-infra.svc.cluster.local').replace(/\/$/, '');
    this.apiKey = config.api_key || '';
  }

  /**
   * Internal wrapper for fetch requests to Lago.
   */
  private async request(method: string, path: string, body?: any): Promise<any> {
    const url = `${this.url}/api/v1/${path.replace(/^\/+/, '')}`;
    const options: RequestInit = {
      method,
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
      },
    };

    if (body) {
      options.body = JSON.stringify(body);
    }

    const response = await fetch(url, options);

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Lago request failed [${response.status}]: ${errorText}`);
    }

    // Handle empty 200/204 responses
    if (response.status === 204 || response.headers.get('content-length') === '0') {
      return null;
    }

    return await response.json();
  }

  /**
   * Create a new customer in Lago.
   */
  public async createCustomer(customerData: Record<string, any>): Promise<any> {
    const data = await this.request('POST', 'customers', { customer: customerData });
    return data.customer;
  }

  /**
   * Create a subscription for a customer.
   */
  public async createSubscription(subscriptionData: Record<string, any>): Promise<any> {
    const data = await this.request('POST', 'subscriptions', { subscription: subscriptionData });
    return data.subscription;
  }

  /**
   * Send a usage event to Lago for a specific customer/subscription.
   */
  public async addUsage(eventData: Record<string, any>): Promise<any> {
    const data = await this.request('POST', 'events', { event: eventData });
    return data.event;
  }

  /**
   * List invoices for a specific customer.
   */
  public async invoices(externalCustomerId: string): Promise<any[]> {
    const data = await this.request('GET', `invoices?external_customer_id=${encodeURIComponent(externalCustomerId)}`);
    return data.invoices || [];
  }

  /**
   * Basic Lago health check.
   */
  public async health(): Promise<boolean> {
    try {
      if (!this.apiKey) return false;
      const response = await fetch(`${this.url}/api/v1/plans`, {
        headers: { 'Authorization': `Bearer ${this.apiKey}` },
      });
      return response.ok;
    } catch {
      return false;
    }
  }
}

File: dotnet/Zirve.Sdk/Billing/BillingManager.cs
using System;
using System.Net.Http;
using System.Net.Http.Json;
using System.Text.Json;
using System.Threading.Tasks;
using Zirve.Sdk.Config;

namespace Zirve.Sdk.Billing;

public class BillingManager
{
    private readonly HttpClient _httpClient;
    private readonly string _url;

    public BillingManager(HttpClient httpClient, ConfigManager configManager)
    {
        _httpClient = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
        
        var cfg = configManager.Module("billing");
        _url = (cfg.GetValueOrDefault("url", "http://lago-api.zirve-infra.svc.cluster.local") ?? "").TrimEnd('/') + "/api/v1";
        
        var key = cfg.GetValueOrDefault("api_key", "");
        if (!string.IsNullOrEmpty(key))
        {
            _httpClient.DefaultRequestHeaders.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", key);
        }
    }

    private async Task<JsonElement> RequestAsync(HttpMethod method, string path, object? body = null)
    {
        var request = new HttpRequestMessage(method, $"{_url}/{path.TrimStart('/')}");
        
        if (body != null)
        {
            var json = body is string s ? s : JsonSerializer.Serialize(body);
            request.Content = new StringContent(json, System.Text.Encoding.UTF8, "application/json");
        }

        var response = await _httpClient.SendAsync(request);
        
        if (!response.IsSuccessStatusCode && response.StatusCode != System.Net.HttpStatusCode.NotFound)
        {
            var err = await response.Content.ReadAsStringAsync();
            throw new Exception($"Lago Billing API Error [{response.StatusCode}]: {err}");
        }

        if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
            return default;

        var jsonStr = await response.Content.ReadAsStringAsync();
        return string.IsNullOrEmpty(jsonStr) ? default : JsonSerializer.Deserialize<JsonElement>(jsonStr);
    }

    public async Task<JsonElement> CreateCustomerAsync(object customer)
    {
        return await RequestAsync(HttpMethod.Post, "customers", new { customer });
    }

    public async Task<JsonElement> CreateSubscriptionAsync(object subscription)
    {
        return await RequestAsync(HttpMethod.Post, "subscriptions", new { subscription });
    }

    public async Task<JsonElement> AddEventAsync(object evt)
    {
        return await RequestAsync(HttpMethod.Post, "events", new { @event = evt });
    }

    public async Task<JsonElement> GetInvoiceAsync(string id)
    {
        return await RequestAsync(HttpMethod.Get, $"invoices/{Uri.EscapeDataString(id)}");
    }

    public async Task<bool> HealthAsync()
    {
        try
        {
            var response = await _httpClient.GetAsync($"{_url}/organizations");
            // If we get 401 Unauthorized it means API is up but token is bad, which still passes health check for reachability
            return response.IsSuccessStatusCode || response.StatusCode == System.Net.HttpStatusCode.Unauthorized;
        }
        catch
        {
            return false;
        }
    }
}
